wallet-new/src/Cardano/Wallet/Kernel/DB/Util/AcidState.hs

```diff
@@ -25,27 +26,41 @@ import           Cardano.Wallet.Kernel.DB.Util.IxSet (Indexable, IxSet)
 import qualified Cardano.Wallet.Kernel.DB.Util.IxSet as IxSet
 
 {-------------------------------------------------------------------------------
-  Acid-state updates with support for errors (and zooming, see below)
+  Acid-state updates with support for errors (and zooming, see below).
+  NOTA BENE: Your acid-state 'Update' queries should be composed by one
+  @and only one@ \"runUpdate'\", as each call to \"runUpdate'\" will return
+  the modified copy of the state 'st' alongside with the final result, so
+  in presence of multiple calls it would be very error prone: one wrong return
+  and an 'Update' query could return a stale version of the DB.
 -------------------------------------------------------------------------------}
 
 type Update' st e = StateT st (Except e)
 
-runUpdate' :: forall e st a. Update' st e a -> Update st (Either e a)
+runUpdate' :: forall e st a. Update' st e a -> Update st (Either e (st, a))
 runUpdate' upd = do
     st <- get
     case upd' st of
       Left  e        -> return (Left e)
-      Right (a, st') -> put st' >> return (Right a)
+      Right (a, st') -> put st' >> return (Right (st, a))
   where
     upd' :: st -> Either e (a, st)
     upd' = runExcept . runStateT upd
 
 runUpdateNoErrors :: Update' st Void a -> Update st a
-runUpdateNoErrors = fmap mustBeRight . runUpdate'
+runUpdateNoErrors = fmap (snd . mustBeRight) . runUpdate'
 
 mapUpdateErrors :: (e -> e') -> Update' st e a -> Update' st e' a
 mapUpdateErrors f upd = StateT $ withExcept f . runStateT upd
 
+-- | Like \"runUpdate'\", but it discards the DB after running the query.
+-- Use this function sparingly only when you are sure you won't need the
+-- DB snapshot afterwards. If you find yourself willing to re-access the DB
+-- after you ran the 'Update', it means you didn't need this function in the
+-- first place.
```

Comment:
I find the sentence "If you find yourself ... first place" somewhat confusingly worded.

Reply:
Will reword.